### ISSUE TITLE
kprobe_pwru.c:set_tuple() reads wrong L4 for ESP on some kprobes

### ISSUE THREAD

**Issue by jschwinger233** [cilium/pwru#205]
## TL;DR

Current `set_tuple()` reads `l4_off` from `skb->transport_header` and then get `struct tcphdr *tcp` based on that:

https://github.com/cilium/pwru/blob/f470db0673c3756faaabab68a0e705b534f69482/bpf/kprobe_pwru.c#L205-L208

https://github.com/cilium/pwru/blob/f470db0673c3756faaabab68a0e705b534f69482/bpf/kprobe_pwru.c#L227-L228

But on some kprobes such as `xfrm_rcv_cb` and `xfrm6_rcv_cb`, `skb->head + skb->transport_header` points to the same position as `skb->head + skb->network_header`, causing wrong L4 tuple settings therefore.

## Long story

I was observing cilium with IPsec using pwru, while keeping curling from pod1@host1 to pod2@host2 using IPv6.

The pwru command below filters tcp syn traffic on host2 (receiving host):

```
$ s pwru --output-tuple --output-meta --filter-dst-port 80  --filter-dst-ip fd00:10:244:1::cf33 --filter-pcap 'ip6[40+13] & 0x3f = 0x2' --filter-netns 4026533408 
2023/07/01 12:43:37 Per cpu buffer size: 4096 bytes
2023/07/01 12:43:37 Attaching kprobes (via kprobe-multi)...
1498 / 1498 [-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------] 100.00% ? p/s
2023/07/01 12:43:37 Attached (ignored 0)
2023/07/01 12:43:37 Listening for events..
               SKB    CPU          PROCESS                     FUNC
0xffff98ae4743eae8      0           [curl]              xfrm_rcv_cb netns=4026533408 mark=0xd00 ifindex=4 proto=dd86 mtu=1500 len=80 [fd00:10:244::834b]:24589->[fd00:10:244:1::cf33]:62723(tcp)
0xffff98ae4743eae8      0           [curl]             xfrm6_rcv_cb netns=4026533408 mark=0xd00 ifindex=4 proto=dd86 mtu=1500 len=80 [fd00:10:244::834b]:24589->[fd00:10:244:1::cf33]:62723(tcp)
0xffff98ae4743eae8      0           [curl]        gro_cells_receive netns=4026533408 mark=0xd00 ifindex=4 proto=dd86 mtu=1500 len=80 [fd00:10:244::834b]:24589->[fd00:10:244:1::cf33]:62723(tcp)
0xffff98ae4743eae8      0           [curl]         napi_gro_receive netns=4026533408 mark=0xd00 ifindex=4 proto=dd86 mtu=1500 len=80 [fd00:10:244::834b]:24589->[fd00:10:244:1::cf33]:62723(tcp)
0xffff98ae4743eae8      0           [curl]          dev_gro_receive netns=4026533408 mark=0xd00 ifindex=4 proto=dd86 mtu=1500 len=80 [fd00:10:244::834b]:24589->[fd00:10:244:1::cf33]:62723(tcp)
0xffff98ae4743eae8      0           [curl]         ipv6_gro_receive netns=4026533408 mark=0xd00 ifindex=4 proto=dd86 mtu=1500 len=80 [fd00:10:244::834b]:24589->[fd00:10:244:1::cf33]:62723(tcp)
0xffff98ae4743eae8      0           [curl]         tcp6_gro_receive netns=4026533408 mark=0xd00 ifindex=4 proto=dd86 mtu=1500 len=80 [fd00:10:244::834b]:37360->[fd00:10:244:1::cf33]:80(tcp)
0xffff98ae4743eae8      0           [curl] __skb_gro_checksum_complete netns=4026533408 mark=0xd00 ifindex=4 proto=dd86 mtu=1500 len=80 [fd00:10:244::834b]:37360->[fd00:10:244:1::cf33]:80(tcp)
0xffff98ae4743eae8      0           [curl]          tcp_gro_receive netns=4026533408 mark=0xd00 ifindex=4 proto=dd86 mtu=1500 len=80 [fd00:10:244::834b]:37360->[fd00:10:244:1::cf33]:80(tcp)
0xffff98ae4743eae8      0           [curl]   skb_defer_rx_timestamp netns=4026533408 mark=0xd00 ifindex=4 proto=dd86 mtu=1500 len=80 [fd00:10:244::834b]:37360->[fd00:10:244:1::cf33]:80(tcp)
0xffff98ae4743eae8      0           [curl]             tcf_classify netns=4026533408 mark=0xd00 ifindex=4 proto=dd86 mtu=1500 len=80 [fd00:10:244::834b]:37360->[fd00:10:244:1::cf33]:80(tcp)
0xffff98ae4743eae8      0           [curl]      skb_ensure_writable netns=4026533408 mark=0xd00 ifindex=4 proto=dd86 mtu=1500 len=94 [fd00:10:244::834b]:37360->[fd00:10:244:1::cf33]:80(tcp)
0xffff98ae4743eae8      0           [curl]      skb_ensure_writable netns=4026533408 mark=0xd00 ifindex=4 proto=dd86 mtu=1500 len=94 [fd00:10:244::834b]:37360->[fd00:10:244:1::cf33]:80(tcp)
0xffff98ae4743eae8      0           [curl]      skb_ensure_writable netns=4026533408 mark=0xd00 ifindex=4 proto=dd86 mtu=1500 len=94 [fd00:10:244::834b]:37360->[fd00:10:244:1::cf33]:80(tcp)
0xffff98ae4743eae8      0           [curl]      skb_ensure_writable netns=4026533408 mark=0x0 ifindex=4 proto=dd86 mtu=1500 len=94 [fd00:10:244::834b]:37360->[fd00:10:244:1::cf33]:80(tcp)
0xffff98ae4743eae8      0           [curl]      skb_ensure_writable netns=4026533408 mark=0x0 ifindex=4 proto=dd86 mtu=1500 len=94 [fd00:10:244::834b]:37360->[fd00:10:244:1::cf33]:80(tcp)
0xffff98ae4743eae8      0           [curl]      skb_ensure_writable netns=4026533408 mark=0x0 ifindex=4 proto=dd86 mtu=1500 len=94 [fd00:10:244::834b]:37360->[fd00:10:244:1::cf33]:80(tcp)
0xffff98ae4743eae8      0           [curl]          skb_do_redirect netns=4026533408 mark=0x0 ifindex=4 proto=dd86 mtu=1500 len=94 [fd00:10:244::834b]:37360->[fd00:10:244:1::cf33]:80(tcp)
0xffff98ae4743eae8      0           [curl]           __bpf_redirect netns=4026533408 mark=0x0 ifindex=4 proto=dd86 mtu=1500 len=94 [fd00:10:244::834b]:37360->[fd00:10:244:1::cf33]:80(tcp)
0xffff98ae4743eae8      0           [curl]         __dev_queue_xmit netns=4026533408 mark=0x0 ifindex=11 proto=dd86 mtu=1500 len=94 [fd00:10:244::834b]:37360->[fd00:10:244:1::cf33]:80(tcp)
0xffff98ae4743eae8      0           [curl]       qdisc_pkt_len_init netns=4026533408 mark=0x0 ifindex=11 proto=dd86 mtu=1500 len=94 [fd00:10:244::834b]:37360->[fd00:10:244:1::cf33]:80(tcp)
0xffff98ae4743eae8      0           [curl]      netdev_core_pick_tx netns=4026533408 mark=0x0 ifindex=11 proto=dd86 mtu=1500 len=94 [fd00:10:244::834b]:37360->[fd00:10:244:1::cf33]:80(tcp)
0xffff98ae4743eae8      0           [curl]        validate_xmit_skb netns=4026533408 mark=0x0 ifindex=11 proto=dd86 mtu=1500 len=94 [fd00:10:244::834b]:37360->[fd00:10:244:1::cf33]:80(tcp)
0xffff98ae4743eae8      0           [curl]       netif_skb_features netns=4026533408 mark=0x0 ifindex=11 proto=dd86 mtu=1500 len=94 [fd00:10:244::834b]:37360->[fd00:10:244:1::cf33]:80(tcp)
0xffff98ae4743eae8      0           [curl]  passthru_features_check netns=4026533408 mark=0x0 ifindex=11 proto=dd86 mtu=1500 len=94 [fd00:10:244::834b]:37360->[fd00:10:244:1::cf33]:80(tcp)
0xffff98ae4743eae8      0           [curl]     skb_network_protocol netns=4026533408 mark=0x0 ifindex=11 proto=dd86 mtu=1500 len=94 [fd00:10:244::834b]:37360->[fd00:10:244:1::cf33]:80(tcp)
0xffff98ae4743eae8      0           [curl]       validate_xmit_xfrm netns=4026533408 mark=0x0 ifindex=11 proto=dd86 mtu=1500 len=94 [fd00:10:244::834b]:37360->[fd00:10:244:1::cf33]:80(tcp)
0xffff98ae4743eae8      0           [curl]      dev_hard_start_xmit netns=4026533408 mark=0x0 ifindex=11 proto=dd86 mtu=1500 len=94 [fd00:10:244::834b]:37360->[fd00:10:244:1::cf33]:80(tcp)
0xffff98ae4743eae8      0           [curl]   skb_clone_tx_timestamp netns=4026533408 mark=0x0 ifindex=11 proto=dd86 mtu=1500 len=94 [fd00:10:244::834b]:37360->[fd00:10:244:1::cf33]:80(tcp)
0xffff98ae4743eae8      0           [curl]        __dev_forward_skb netns=4026533408 mark=0x0 ifindex=11 proto=dd86 mtu=1500 len=94 [fd00:10:244::834b]:37360->[fd00:10:244:1::cf33]:80(tcp)
0xffff98ae4743eae8      0           [curl]       __dev_forward_skb2 netns=4026533408 mark=0x0 ifindex=11 proto=dd86 mtu=1500 len=94 [fd00:10:244::834b]:37360->[fd00:10:244:1::cf33]:80(tcp)
0xffff98ae4743eae8      0           [curl]         skb_scrub_packet netns=4026533408 mark=0x0 ifindex=11 proto=dd86 mtu=1500 len=94 [fd00:10:244::834b]:37360->[fd00:10:244:1::cf33]:80(tcp)
0xffff98ae4743eae8      0           [curl]           eth_type_trans netns=4026533408 mark=0x0 ifindex=11 proto=dd86 mtu=1500 len=94 [fd00:10:244::834b]:37360->[fd00:10:244:1::cf33]:80(tcp)
```

Please notice the first 6 lines of output, which printed skb with dst port 62723, this is not what we want as we specified `--filter-dst-port 80`.

## Investigation

Let's run that pwru command again but with `output-skb`:

```
0xffff98ae4743dae8      0           [curl]              xfrm_rcv_cb netns=4026533408 mark=0xd00 ifindex=4 proto=dd86 mtu=1500 len=80 [fd00:10:244::834b]:24584->[fd00:10:244:1::cf33]:55600(tcp)
(struct sk_buff){
 (union){
  (struct){
   .prev = (struct sk_buff *)0x00000000cd64749d,
   (union){
    .dev = (struct net_device *)0x000000008db904fc,
    .dev_scratch = (long unsigned int)18446630474028769280,
   },
  },
  .rbnode = (struct rb_node){
   .rb_right = (struct rb_node *)0x00000000cd64749d,
   .rb_left = (struct rb_node *)0x000000008db904fc,
  },
  .list = (struct list_head){
   .prev = (struct list_head *)0x00000000cd64749d,
  },
 },
 .cb = (char[])[
  4,
 ],
 (union){
  (struct){
   ._skb_refdst = (long unsigned int)18446630476639842048,
  },
  .tcp_tsorted_anchor = (struct list_head){
   .next = (struct list_head *)0x000000003ed00c73,
  },
  ._sk_redir = (long unsigned int)18446630476639842048,
 },
 ._nfct = (long unsigned int)7,
 .len = (unsigned int)80,
 .mac_len = (__u16)14,
 .fclone = (__u8)0x2,
 .active_extensions = (__u8)2,
 (union){
  (struct){
   .ooo_okay = (__u8)0x1,
   .l4_hash = (__u8)0x1,
   .csum_valid = (__u8)0x1,
   .tc_at_ingress = (__u8)0x1,
   .slow_gro = (__u8)0x1,
   (union){
    .csum = (__wsum)1048856,
    (struct){
     .csum_start = (__u16)280,
     .csum_offset = (__u16)16,
    },
   },
   .skb_iif = (int)4,
   .hash = (__u32)4162138329,
   (union){
    .mark = (__u32)3328,
    .reserved_tailroom = (__u32)3328,
   },
   (union){
    .inner_protocol = (__be16)22629,
    .inner_ipproto = (__u8)101,
   },
   .inner_transport_header = (__u16)224,
   .inner_network_header = (__u16)184,
   .inner_mac_header = (__u16)170,
   .protocol = (__be16)56710,
   .transport_header = (__u16)240,
   .network_header = (__u16)240,
   .mac_header = (__u16)226,
  },
  .headers = (struct){
   .ooo_okay = (__u8)0x1,
   .l4_hash = (__u8)0x1,
   .csum_valid = (__u8)0x1,
   .tc_at_ingress = (__u8)0x1,
   .slow_gro = (__u8)0x1,
   (union){
    .csum = (__wsum)1048856,
    (struct){
     .csum_start = (__u16)280,
     .csum_offset = (__u16)16,
    },
   },
   .skb_iif = (int)4,
   .hash = (__u32)4162138329,
   (union){
    .mark = (__u32)3328,
    .reserved_tailroom = (__u32)3328,
   },
```

The `transport_header` and `network_header` have the same value `240`! 

That explains everything: the `filter_pcap()` works as expected because it calculates L4 using L3 header pointer, which is accurate; `set_output()` on the other hand relies on `skb->transport_header`, which isn't set properly for those kprobes.

To explain how we get a strange dest port 62723, let's inspect the packet:

![image](https://github.com/cilium/pwru/assets/6435624/ca6388ea-a97a-418a-8640-218fa41baba9)

In output_tuple(), `struct tcphdr *tcp` points to L3 header wrongly, as I highlighted in the picture above. Then `tcp->dport` would get the 3rd and 4th bytes: `f5 03`, which is exactly 62723 in hex: 

```
>>> 0xf503
62723
```

In conclusion, pcap filter counts on the L3 header only and works perfect to filter the packet, but tuple output setting doesn't make it right.

**Comment by brb** [comment]
Nice report, thanks. If I understood correctly, the easy fix would be to always derive the L4 hdr from the L3 hdr?

**Comment by jschwinger233** [comment]
@brb Precisely!
